### PR TITLE
planner: refactor infoschema predicate extractors

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref #50305

Problem Summary:

Previously, we have a few usages like `ex.Filter("partition_name", pi.Name.L)` which is easy to misuse.

### What changed and how does it work?

- Add `HasXXX` method for each extractors, remove `Filter`s.
- Extract similar code `ListSchemasAndTables` to `InfoSchemaBaseExtractor`.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/infoschema_reader.go [hunk 1]
```diff
@@ -1748,10 +1741,10 @@ func (e *memtableRetriever) setDataForMetricTables() {
 	e.rows = rows
 }
 
-func keyColumnUsageInTable(schema model.CIStr, table *model.TableInfo, extractor *plannercore.InfoSchemaKeyColumnUsageExtractor) [][]types.Datum {
+func keyColumnUsageInTable(schema model.CIStr, table *model.TableInfo, ex *plannercore.InfoSchemaKeyColumnUsageExtractor) [][]types.Datum {
 	var rows [][]types.Datum
 	if table.PKIsHandle {
```
Comment (wjhuang2016) [inline pkg/executor/infoschema_reader.go:1746]: Merge L1746 and L1747

Comment (tangenta) [reply pkg/executor/infoschema_reader.go]: https://github.com/pingcap/tidb/pull/55502/commits/569fc30d7aec000fa7a607c83db820d97cf0f4d3

File: pkg/executor/infoschema_reader.go [hunk 2]
```diff
@@ -1792,7 +1783,7 @@ func keyColumnUsageInTable(schema model.CIStr, table *model.TableInfo, extractor
 			continue
 		}
 
-		if extractor != nil && extractor.Filter("constraint_name", filterIdxName) {
+		if !ex.HasConstraint(filterIdxName) {
```
Comment (tiancaiamao) [inline pkg/executor/infoschema_reader.go:1786]: This makes no difference.
People forge to write the `extractor.Filter(` ... they can also forget to write `HasXXX`?

**Review by tiancaiamao — APPROVED** [review]

File: pkg/executor/infoschema_reader.go (hunk 2 above)
Comment (tangenta) [reply pkg/executor/infoschema_reader.go:1786]: Just trying to avoid using string literal here

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**